File: tooling/check-company-updates.py
#!/usr/bin/env python3
"""
Check recent product updates from tracked companies in context/companies.md

This script:
1. Parses context/companies.md to find companies and their primary sources
2. Checks RSS feeds from company blogs (using feedparser, like check-recent-posts.py)
3. Outputs recent updates in markdown format

Note: Changelog scraping would require Puppeteer and is more complex.
For now, this focuses on RSS feeds which are more reliable.
"""

import re
import sys
import json
import argparse
from datetime import datetime, timedelta
from pathlib import Path
from urllib.parse import urlparse

try:
    import feedparser
except ImportError:
    print("Error: feedparser not installed. Install with: pip install feedparser", file=sys.stderr)
    sys.exit(1)

try:
    import requests
    from bs4 import BeautifulSoup
except ImportError:
    print("Error: requests or beautifulsoup4 not installed. Install with: pip install requests beautifulsoup4", file=sys.stderr)
    sys.exit(1)

# Add project root to path
PROJECT_ROOT = Path(__file__).parent.parent
sys.path.insert(0, str(PROJECT_ROOT))

def parse_companies_file(companies_file_path):
    """Parse companies.md to extract company info including blogs and changelogs."""
    companies = []
    
    # Known RSS feeds (can be expanded)
    known_feeds = {
        'LangChain / LangSmith': ['https://blog.langchain.dev/feed'],
        'GitHub': ['https://github.blog/feed/'],
    }
    
    with open(companies_file_path, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Split by company sections (## Name)
    sections = re.split(r'\n## ', content)
    
    for section in sections[1:]:  # Skip header
        lines = section.split('\n')
        name = lines[0].strip()
        
        company = {
            'name': name,
            'blogs': [],
            'rss_feeds': [],
            'changelogs': [],
            'category': None,
        }
        
        # Check for known RSS feeds
        if name in known_feeds:
            company['rss_feeds'] = known_feeds[name]
        
        # Extract category
        category_match = re.search(r'\*\*Category:\*\* (.+)', section)
        if category_match:
            company['category'] = category_match.group(1)
        
        # Extract primary sources
        in_primary_sources = False
        for line in lines:
            if '**Primary sources:**' in line:
                in_primary_sources = True
                continue
            if in_primary_sources:
                # Stop at next section or empty line after sources
                if line.strip() == '' and company['blogs']:
                    break
                if line.startswith('---'):
                    break
                
                # Extract URLs
                url_matches = re.findall(r'https?://[^\s\)]+', line)
                for url in url_matches:
                    clean_url = url.rstrip(')').rstrip(',')
                    
                    # Check if it's an RSS feed
                    if any(x in clean_url.lower() for x in ['/feed', '/rss', '/atom']):
                        company['rss_feeds'].append(clean_url)
                    # Categorize URLs
                    elif 'changelog' in clean_url.lower() or 'release-notes' in clean_url.lower() or 'docs/changelog' in clean_url.lower():
                        company['changelogs'].append(clean_url)
                    elif 'blog' in clean_url.lower() or 'news' in clean_url.lower() or 'updates' in clean_url.lower():
                        company['blogs'].append(clean_url)
                    elif 'twitter.com' in clean_url or 'x.com' in clean_url:
                        pass  # Skip Twitter for now
                    elif 'docs' not in clean_url.lower() or 'changelog' in clean_url.lower():
                        # Default to blog if not clearly a changelog
                        company['blogs'].append(clean_url)
        
        if company['blogs'] or company['changelogs'] or company['rss_feeds']:
            companies.append(company)
    
    return companies

def try_find_rss_feed(blog_url):
    """Try to find RSS feed URL from blog homepage."""
    common_rss_paths = [
        '/feed', '/feed.xml', '/rss', '/rss.xml', '/atom.xml', '/index.xml',
        '/blog/feed', '/blog/rss', '/blog/atom.xml',
        '/feeds/posts/default',  # Blogger
        '/feed/rss', '/feed/atom',
    ]
    
    try:
        response = requests.get(blog_url, timeout=10, headers={'User-Agent': 'Mozilla/5.0'})
        soup = BeautifulSoup(response.content, 'html.parser')
        
        # Look for RSS link in HTML - try multiple selectors
        rss_link = (soup.find('link', {'type': 'application/rss+xml'}) or 
                   soup.find('link', {'type': 'application/atom+xml'}) or
                   soup.find('link', {'type': 'text/xml'}) or
                   soup.find('a', href=re.compile(r'feed|rss|atom', re.I)))
        
        if rss_link:
            href = rss_link.get('href') or rss_link.get('href')
            if href:
                if href.startswith('http'):
                    return href
                else:
                    # Relative URL
                    parsed = urlparse(blog_url)
                    return f"{parsed.scheme}://{parsed.netloc}{href}"
        
        # Try common paths - check both HEAD and GET
        parsed = urlparse(blog_url)
        base_url = f"{parsed.scheme}://{parsed.netloc}"
        
        # Also try with blog path if blog_url has a path
        if parsed.path and parsed.path != '/':
            base_paths = [parsed.path, '/']
        else:
            base_paths = ['/']
        
        for base_path in base_paths:
            for path in common_rss_paths:
                test_url = base_url + base_path.rstrip('/') + path
                try:
                    # Try HEAD first
                    test_response = requests.head(test_url, timeout=5, allow_redirects=True)
                    if test_response.status_code == 200:
                        content_type = test_response.headers.get('content-type', '').lower()
                        if any(x in content_type for x in ['xml', 'rss', 'atom']):
                            return test_url
                    
                    # If HEAD doesn't work, try GET and check content
                    test_response = requests.get(test_url, timeout=5, allow_redirects=True)
                    if test_response.status_code == 200:
                        content = test_response.text[:500].lower()
                        if any(x in content for x in ['<rss', '<feed', '<?xml', 'atom']):
                            return test_url
                except:
                    continue
        
    except Exception as e:
        pass
    
    return None

def check_rss_feed(feed_url, days_back=7):
    """Check RSS feed for recent posts."""
    try:
        # Use requests to fetch the feed
        try:
            response = requests.get(feed_url, timeout=10, headers={'User-Agent': 'Mozilla/5.0'}, verify=True)
        except requests.exceptions.SSLError:
            import urllib3
            urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
            response = requests.get(feed_url, timeout=10, headers={'User-Agent': 'Mozilla/5.0'}, verify=False)
        
        response.raise_for_status()
        
        # Parse the feed content with feedparser
        feed = feedparser.parse(response.content)
        
        if feed.bozo and feed.bozo_exception:
            if 'not well-formed' not in str(feed.bozo_exception).lower():
                return None, f"RSS feed error: {feed.bozo_exception}"
        
        cutoff_date = datetime.now() - timedelta(days=days_back)
        recent_posts = []
        
        for entry in feed.entries[:15]:  # Check last 15 entries
            # Try to parse date
            pub_date = None
            if hasattr(entry, 'published_parsed') and entry.published_parsed:
                pub_date = datetime(*entry.published_parsed[:6])
            elif hasattr(entry, 'updated_parsed') and entry.updated_parsed:
                pub_date = datetime(*entry.updated_parsed[:6])
            
            if pub_date and pub_date >= cutoff_date:
                post = {
                    'title': entry.get('title', 'Untitled'),
                    'link': entry.get('link', ''),
                    'published': pub_date.isoformat() if pub_date else None,
                    'summary': entry.get('summary', '')[:500],  # First 500 chars
                }
                recent_posts.append(post)
            elif not pub_date:
                # If no date, include it anyway (might be recent)
                post = {
                    'title': entry.get('title', 'Untitled'),
                    'link': entry.get('link', ''),
                    'published': None,
                    'summary': entry.get('summary', '')[:500],
                }
                recent_posts.append(post)
        
        return recent_posts, None
        
    except requests.exceptions.RequestException as e:
        return None, f"Error fetching RSS feed: {str(e)}"
    except Exception as e:
        return None, f"Error checking RSS feed: {str(e)}"

def check_company_updates(company, days_back=7):
    """Check recent updates for a company."""
    updates = []
    errors = []
    
    # Check known RSS feeds first
    for rss_feed in company.get('rss_feeds', []):
        posts, error = check_rss_feed(rss_feed, days_back)
        if error:
            errors.append(f"{rss_feed}: {error}")
        elif posts:
            for post in posts:
                updates.append({
                    **post,
                    'source': 'rss',
                    'source_url': rss_feed,
                })
    
    # Check blogs via RSS discovery
    for blog_url in company['blogs']:
        # Try to find RSS feed
        rss_feed = try_find_rss_feed(blog_url)
        if rss_feed:
            posts, error = check_rss_feed(rss_feed, days_back)
            if error:
                errors.append(f"{blog_url}: {error}")
            elif posts:
                for post in posts:
                    updates.append({
                        **post,
                        'source': 'blog',
                        'source_url': blog_url,
                    })
        else:
            # Only report as error if we don't have a known feed
            if not company.get('rss_feeds'):
                errors.append(f"{blog_url}: No RSS feed found")
    
    # Note: Changelog scraping would require Puppeteer
    # For now, we skip changelogs and focus on RSS feeds
    
    return updates, errors

def format_output(results, output_format='json'):
    """Format results for output."""
    if output_format == 'json':
        return json.dumps(results, indent=2, ensure_ascii=False)
    elif output_format == 'markdown':
        output = "# Recent Company Updates\n\n"
        for result in results:
            if result['updates']:
                output += f"## {result['name']}\n"
                if result['category']:
                    output += f"*Category: {result['category']}*\n\n"
                
                for update in result['updates']:
                    output += f"### {update['title']}\n"
                    output += f"**Link:** {update['link']}\n"
                    if update['published']:
                        output += f"**Published:** {update['published']}\n"
                    output += f"**Source:** {update['source']} ({update['source_url']})\n"
                    if update['summary']:
                        output += f"**Summary:** {update['summary']}\n"
                    output += "\n"
        return output
    else:
        return str(results)

def main():
    parser = argparse.ArgumentParser(description='Check recent updates from tracked companies')
    parser.add_argument('--days', type=int, default=7, help='Number of days back to check (default: 7)')
    parser.add_argument('--format', choices=['json', 'markdown'], default='markdown', help='Output format')
    parser.add_argument('--companies-file', type=str, 
                       default=str(PROJECT_ROOT / 'context' / 'companies.md'),
                       help='Path to companies.md file')
    
    args = parser.parse_args()
    
    # Parse companies file
    companies = parse_companies_file(args.companies_file)
    
    if not companies:
        print("No companies with blogs or changelogs found.", file=sys.stderr)
        return 1
    
    print(f"Checking company updates from last {args.days} days...\n")
    print(f"Found {len(companies)} companies with sources\n")
    
    # Check recent updates
    results = []
    for company in companies:
        print(f"Checking {company['name']}...")
        updates, errors = check_company_updates(company, days_back=args.days)
        
        results.append({
            'name': company['name'],
            'category': company['category'],
            'updates': updates,
            'errors': errors,
        })
        
        if updates:
            print(f"  ✓ Found {len(updates)} updates")
        if errors:
            print(f"  ⚠ {len(errors)} errors:")
            for error in errors[:2]:  # Show first 2 errors
                print(f"    - {error}")
    
    # Output results
    print("\n" + "="*60 + "\n")
    print(format_output(results, args.format))
    
    return 0

if __name__ == '__main__':
    sys.exit(main())

File: tooling/check-company-updates.js
#!/usr/bin/env node
/**
 * @deprecated This script has been replaced by the modular version.
 * Please use: node scripts/check-company-updates.js
 * 
 * Check recent product updates from tracked companies
 * 
 * This script:
 * 1. Parses context/companies.md to find companies and their primary sources
 * 2. Checks RSS feeds from company blogs
 * 3. Scrapes changelog pages for recent updates
 * 4. Optionally checks for news mentions
 * 5. Outputs recent updates in markdown format
 */

const puppeteer = require('puppeteer');
const fs = require('fs');
const path = require('path');
const { URL } = require('url');

const PROJECT_ROOT = path.join(__dirname, '..');
const COMPANIES_FILE = path.join(PROJECT_ROOT, 'context', 'companies.md');

// Number of days to look back
const DEFAULT_DAYS_BACK = 14;

function parseCompaniesFile() {
  const content = fs.readFileSync(COMPANIES_FILE, 'utf-8');
  const companies = [];
  
  // Split by company sections (## Company Name)
  const sections = content.split(/\n## /);
  
  for (const section of sections.slice(1)) {
    const lines = section.split('\n');
    const name = lines[0].trim();
    
    const company = {
      name,
      blogs: [],
      changelogs: [],
      twitter: null,
      category: null,
    };
    
    // Extract category
    const categoryMatch = section.match(/\*\*Category:\*\* (.+)/);
    if (categoryMatch) {
      company.category = categoryMatch[1];
    }
    
    // Extract primary sources
    let inPrimarySources = false;
    for (const line of lines) {
      if (line.includes('**Primary sources:**')) {
        inPrimarySources = true;
        continue;
      }
      if (inPrimarySources) {
        // Stop at next section or empty line after sources
        if (line.trim() === '' && company.blogs.length > 0) {
          break;
        }
        if (line.startsWith('---')) {
          break;
        }
        
        // Extract URLs
        const urlMatch = line.match(/https?:\/\/[^\s\)]+/g);
        if (urlMatch) {
          for (const url of urlMatch) {
            const cleanUrl = url.replace(/\)$/, '').replace(/,$/, '');
            
            // Categorize URLs
            if (cleanUrl.includes('blog') || cleanUrl.includes('news') || cleanUrl.includes('updates')) {
              company.blogs.push(cleanUrl);
            } else if (cleanUrl.includes('changelog') || cleanUrl.includes('release-notes') || cleanUrl.includes('docs/changelog')) {
              company.changelogs.push(cleanUrl);
            } else if (cleanUrl.includes('twitter.com') || cleanUrl.includes('x.com')) {
              company.twitter = cleanUrl;
            } else if (!cleanUrl.includes('docs') || cleanUrl.includes('changelog')) {
              // Default to blog if it's not clearly a changelog
              company.blogs.push(cleanUrl);
            }
          }
        }
        
        // Extract Twitter handles
        const twitterMatch = line.match(/@[\w]+/);
        if (twitterMatch) {
          company.twitter = `https://twitter.com/${twitterMatch[0].replace('@', '')}`;
        }
      }
    }
    
    if (company.blogs.length > 0 || company.changelogs.length > 0) {
      companies.push(company);
    }
  }
  
  return companies;
}

async function checkRSSFeed(page, feedUrl, daysBack) {
  try {
    await page.goto(feedUrl, { waitUntil: 'networkidle2', timeout: 15000 });
    
    // Try to parse RSS feed
    const content = await page.content();
    
    // Simple RSS parsing (could be improved with a proper RSS parser)
    const items = [];
    const itemMatches = content.matchAll(/<item>([\s\S]*?)<\/item>/gi);
    
    for (const match of itemMatches) {
      const itemContent = match[1];
      const titleMatch = itemContent.match(/<title>(.*?)<\/title>/i);
      const linkMatch = itemContent.match(/<link>(.*?)<\/link>/i);
      const pubDateMatch = itemContent.match(/<pubDate>(.*?)<\/pubDate>/i);
      const descriptionMatch = itemContent.match(/<description>(.*?)<\/description>/i);
      
      if (titleMatch && linkMatch) {
        const pubDate = pubDateMatch ? new Date(pubDateMatch[1]) : null;
        const cutoffDate = new Date();
        cutoffDate.setDate(cutoffDate.getDate() - daysBack);
        
        if (!pubDate || pubDate >= cutoffDate) {
          items.push({
            title: titleMatch[1].replace(/<!\[CDATA\[(.*?)\]\]>/gi, '$1'),
            link: linkMatch[1].replace(/<!\[CDATA\[(.*?)\]\]>/gi, '$1'),
            published: pubDate ? pubDate.toISOString() : null,
            description: descriptionMatch ? descriptionMatch[1].replace(/<!\[CDATA\[(.*?)\]\]>/gi, '$1').substring(0, 300) : null,
          });
        }
      }
      
      if (items.length >= 10) break; // Limit to 10 most recent
    }
    
    return items;
  } catch (error) {
    console.error(`  ✗ Error checking RSS feed ${feedUrl}: ${error.message}`);
    return [];
  }
}

async function scrapeChangelog(page, changelogUrl, daysBack) {
  try {
    await page.goto(changelogUrl, { waitUntil: 'networkidle2', timeout: 15000 });
    
    // Wait a bit for dynamic content
    await new Promise(resolve => setTimeout(resolve, 2000));
    
    // Try to find recent entries
    const entries = await page.evaluate((daysBack) => {
      const cutoffDate = new Date();
      cutoffDate.setDate(cutoffDate.getDate() - daysBack);
      
      const results = [];
      
      // Look for common changelog patterns
      // Try to find date-based sections or recent entries
      const allElements = document.querySelectorAll('article, .changelog-entry, .release, [class*="changelog"], [class*="release"], h2, h3');
      
      for (const el of allElements) {
        const text = el.textContent || '';
        const dateMatch = text.match(/(\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4})|(January|February|March|April|May|June|July|August|September|October|November|December)\s+\d{1,2},?\s+\d{4}/i);
        
        if (dateMatch) {
          let entryDate;
          try {
            entryDate = new Date(dateMatch[0]);
          } catch (e) {
            continue;
          }
          
          if (entryDate >= cutoffDate) {
            // Find link if available
            const linkEl = el.querySelector('a') || el.closest('a');
            const link = linkEl ? linkEl.href : window.location.href;
            
            // Get title
            const titleEl = el.querySelector('h1, h2, h3, h4, .title, [class*="title"]') || el;
            const title = titleEl.textContent.trim().substring(0, 200);
            
            if (title && title.length > 10) {
              results.push({
                title: title,
                link: link,
                published: entryDate.toISOString(),
                description: el.textContent.substring(0, 300),
              });
            }
          }
        }
      }
      
      // If no date-based entries found, try to get first few entries
      if (results.length === 0) {
        const firstEntries = document.querySelectorAll('article, .changelog-entry, .release, [class*="changelog"]');
        for (let i = 0; i < Math.min(5, firstEntries.length); i++) {
          const el = firstEntries[i];
          const linkEl = el.querySelector('a') || el.closest('a');
          const link = linkEl ? linkEl.href : window.location.href;
          const title = el.textContent.trim().substring(0, 200);
          
          if (title && title.length > 10) {
            results.push({
              title: title,
              link: link,
              published: null,
              description: el.textContent.substring(0, 300),
            });
          }
        }
      }
      
      return results.slice(0, 10); // Limit to 10
    }, daysBack);
    
    return entries;
  } catch (error) {
    console.error(`  ✗ Error scraping changelog ${changelogUrl}: ${error.message}`);
    return [];
  }
}

async function findRSSFeedFromBlog(page, blogUrl) {
  try {
    await page.goto(blogUrl, { waitUntil: 'domcontentloaded', timeout: 10000 });
    
    // Look for RSS link
    const rssLink = await page.evaluate(() => {
      // Check link tags
      const linkTag = document.querySelector('link[type*="rss"], link[type*="atom"]');
      if (linkTag && linkTag.href) {
        return linkTag.href;
      }
      
      // Check anchor tags
      const anchorTag = document.querySelector('a[href*="feed"], a[href*="rss"], a[href*="atom"]');
      if (anchorTag && anchorTag.href) {
        return anchorTag.href;
      }
      
      return null;
    });
    
    if (rssLink) {
      // Make absolute URL if relative
      try {
        return new URL(rssLink, blogUrl).href;
      } catch (e) {
        return rssLink.startsWith('http') ? rssLink : `${blogUrl}${rssLink}`;
      }
    }
    
    // Try common paths
    const commonPaths = ['/feed', '/feed.xml', '/rss', '/rss.xml', '/atom.xml'];
    for (const path of commonPaths) {
      try {
        const testUrl = new URL(path, blogUrl).href;
        const response = await page.goto(testUrl, { waitUntil: 'domcontentloaded', timeout: 5000 });
        const contentType = response.headers()['content-type'] || '';
        if (response.status() === 200 && (contentType.includes('xml') || contentType.includes('rss') || contentType.includes('atom'))) {
          return testUrl;
        }
      } catch (e) {
        continue;
      }
    }
    
    return null;
  } catch (error) {
    return null;
  }
}

async function checkCompanyUpdates(browser, company, daysBack) {
  console.log(`\nChecking ${company.name}...`);
  
  const page = await browser.newPage();
  const updates = [];
  
  try {
    // Check blogs via RSS
    for (const blogUrl of company.blogs) {
      console.log(`  Checking blog: ${blogUrl}`);
      
      // Try to find RSS feed
      const rssFeed = await findRSSFeedFromBlog(page, blogUrl);
      if (rssFeed) {
        console.log(`    Found RSS feed: ${rssFeed}`);
        const items = await checkRSSFeed(page, rssFeed, daysBack);
        updates.push(...items.map(item => ({ ...item, source: 'blog', sourceUrl: blogUrl })));
      } else {
        console.log(`    No RSS feed found, skipping blog`);
      }
      
      await new Promise(resolve => setTimeout(resolve, 1000)); // Be polite
    }
    
    // Check changelogs
    for (const changelogUrl of company.changelogs) {
      console.log(`  Checking changelog: ${changelogUrl}`);
      const entries = await scrapeChangelog(page, changelogUrl, daysBack);
      updates.push(...entries.map(entry => ({ ...entry, source: 'changelog', sourceUrl: changelogUrl })));
      await new Promise(resolve => setTimeout(resolve, 1000));
    }
    
  } catch (error) {
    console.error(`  ✗ Error checking ${company.name}: ${error.message}`);
  } finally {
    await page.close();
  }
  
  return updates;
}

async function main() {
  const args = process.argv.slice(2);
  const daysBack = args.includes('--days') 
    ? parseInt(args[args.indexOf('--days') + 1]) 
    : DEFAULT_DAYS_BACK;
  const format = args.includes('--format') 
    ? args[args.indexOf('--format') + 1] 
    : 'markdown';
  
  console.log(`Checking company updates from last ${daysBack} days...\n`);
  
  const companies = parseCompaniesFile();
  console.log(`Found ${companies.length} companies with sources\n`);
  
  // Use a temporary user data directory to avoid permission issues
  const os = require('os');
  const userDataDir = path.join(os.tmpdir(), 'puppeteer-user-data-' + Date.now());
  
  const launchOptions = {
    headless: true,
    userDataDir: userDataDir, // Use a temporary user data directory
    args: [
      '--no-sandbox',
      '--disable-setuid-sandbox',
      '--disable-dev-shm-usage',
      '--disable-software-rasterizer',
      '--disable-gpu',
      '--disable-crash-reporter',
      '--disable-background-networking',
      '--disable-background-timer-throttling',
      '--disable-backgrounding-occluded-windows',
      '--disable-breakpad',
      '--disable-component-extensions-with-background-pages',
      '--disable-features=TranslateUI',
      '--disable-ipc-flooding-protection',
      '--no-first-run',
      '--no-default-browser-check',
      '--no-pings',
      '--no-zygote',
      // Additional flags for better compatibility in restricted environments
      '--disable-extensions',
      '--disable-plugins',
      '--disable-default-apps',
      '--disable-sync',
      '--metrics-recording-only',
      '--mute-audio',
      '--no-crash-upload',
      '--disable-background-downloads',
      '--disable-client-side-phishing-detection',
      '--disable-hang-monitor',
      '--disable-popup-blocking',
      '--disable-prompt-on-repost',
      '--disable-translate',
      '--disable-web-resources',
      '--safebrowsing-disable-auto-update',
    ],
    // Ignore HTTPS errors (useful for development/local environments)
    ignoreHTTPSErrors: true,
  };
  
  // Use Puppeteer's bundled Chromium for better compatibility
  console.log('Using Puppeteer\'s bundled Chromium\n');
  
  let browser;
  try {
    browser = await puppeteer.launch(launchOptions);
  } catch (error) {
    // Provide helpful error message if launch fails
    console.error('\n✗ Failed to launch browser. This may be due to:');
    console.error('  1. Missing permissions (try running with elevated permissions)');
    console.error('  2. Chrome/Chromium installation issues');
    console.error('  3. Sandbox restrictions');
    console.error(`\nError: ${error.message}\n`);
    console.error('If running in a sandboxed environment, ensure the script has:');
    console.error('  - Network access permissions');
    console.error('  - File system write access to temp directory');
    console.error('  - Ability to launch browser processes\n');
    process.exit(1);
  }
  
  const allUpdates = [];
  
  for (const company of companies) {
    const updates = await checkCompanyUpdates(browser, company, daysBack);
    for (const update of updates) {
      allUpdates.push({
        company: company.name,
        category: company.category,
        ...update
      });
    }
    
    // Small delay between companies
    await new Promise(resolve => setTimeout(resolve, 2000));
  }
  
  await browser.close();
  
  // Clean up user data directory
  try {
    if (fs.existsSync(userDataDir)) {
      fs.rmSync(userDataDir, { recursive: true, force: true });
    }
  } catch (e) {
    // Ignore cleanup errors
  }
  
  // Sort by date (most recent first)
  allUpdates.sort((a, b) => {
    if (!a.published && !b.published) return 0;
    if (!a.published) return 1;
    if (!b.published) return -1;
    return new Date(b.published) - new Date(a.published);
  });
  
  // Output results
  if (format === 'json') {
    console.log(JSON.stringify(allUpdates, null, 2));
  } else {
    console.log('# Recent Company Updates\n');
    
    // Group by company
    const byCompany = {};
    for (const update of allUpdates) {
      if (!byCompany[update.company]) {
        byCompany[update.company] = [];
      }
      byCompany[update.company].push(update);
    }
    
    for (const [companyName, updates] of Object.entries(byCompany)) {
      console.log(`## ${companyName}`);
      if (updates[0].category) {
        console.log(`*Category: ${updates[0].category}*\n`);
      }
      
      for (const update of updates) {
        console.log(`### ${update.title}`);
        console.log(`**Link:** ${update.link}`);
        if (update.published) {
          console.log(`**Published:** ${update.published}`);
        }
        console.log(`**Source:** ${update.source} (${update.sourceUrl})`);
        if (update.description) {
          console.log(`**Summary:** ${update.description}`);
        }
        console.log('');
      }
    }
  }
}

main().catch(console.error);

